### PR TITLE
Update and refine the readme

### PR DESCRIPTION
**Pull request by NickVolynkin**
Preview: https://github.com/NickVolynkin/test-run/blob/update-readme/README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,51 +2,126 @@
...
+```
+
+There's a syntax for single- and multiline lists. (TODO see ./sql-tap/suite.ini)
+For example:
+
+```ini
+single_line = first.test.lua second.test.lua third.test.lua
+multiline = first.test.lua ;
+            second.test.lua ;
+            third.test.lua ;
```
Comment (Totktonada) [inline README.md:24]: AFAIR, semicolons are just comments. They don't affect parsing in this case.

We investigated the questions around precise syntax (multiline options, inline comments) within the patchset for support of Python 3. I can find necessary info, if you want.

File: README.md [hunk 2]
```diff
@@ -2,51 +2,126 @@
...
+
+### General configuration
+
+* `core` — major testing dependency or method
+  Should have one of the following values:
+
+  * `tarantool` — test suite used for functional testing
+  * `app` — TAP-tests, another functional test suite
+  * `unittest` — unit testing test suite
+  * `luatest` — test suite using luatest library
```
Comment (Totktonada) [inline README.md:38]: If you're interested I have a description of all test types [here](https://github.com/tarantool/luatest/issues/191) (under the cut). It reveals one or two main characteristics of a test type (taking care to real usage and informal terminology we have). However it is balanced to be as concise as possible. As for me, current description gives no useful information (sorry).

File: README.md [hunk 3]
```diff
@@ -2,51 +2,126 @@
...
+
+* `description` — test suite description
```
Comment (Totktonada) [inline README.md:40]: AFAIR we have a rule in our developer guidelines to don't write useless comments (which just repeats a variable name, for example). If you want to give a suggestion what to write here (just slug, a few words sumary, verbose description), let's give it. Otherwise it is better to don't comment at all.

File: README.md [hunk 4]
```diff
@@ -2,51 +2,126 @@
...
+* `script` (optional) — shebang file to start tarantool with.
+  Most often, the value is a file in the same directory as `suite.ini`, like `box.lua` or `master.lua`.
```
Comment (Totktonada) [inline README.md:42]: First of all, it is only for `core = tarantool` test suites. It is not optional for them and ignored for all other ones. 'Optional' word is barely applicable here.

'Most often'? I don't know exceptions.

I don't like 'shebang file' term, it sounds a bit strange. How about just `an instance file to start the default server from it (using tarantoolctl); applicable for `core = tarantool` test suites'.

File: README.md [hunk 5]
```diff
@@ -2,51 +2,126 @@
...
+  
+* `config` — name of a test configuration file, for example, `engine.cfg`.
+  TODO: what's a test configuration file?
```
Comment (Totktonada) [inline README.md:45]: Parametrization. Ability to run a test in several configurations. Say, for memtx and vinyl.

File: README.md [hunk 6]
```diff
@@ -2,51 +2,126 @@
...
-executed. In the local env there's object `self`, that's `Test` object. If test
-must be skipped - you must put `self.skip = 1` in this file. Next,
-`.test(.lua|.py)?` is executed and file `.reject` is created, then `.reject` is
-compared with `.result`. If something differs, then 15 last string of this diff
-file are printed and `.reject` file is saving in the `<vardir>/rejects/<suite>`
-subfolder given in options or set localy as `var/rejects/<suite>` by default.
-If not, then `.reject` file is deleted.
+## Test suite configuration
+
+Test suite is a bunch of tests located in a subdirectory (recursively) with a `suite.ini`
```
Comment (Totktonada) [inline README.md:7]: TBH, the recursive search was never used (at least now at last four years) and I doubt that it works (even if it initially did). I would not mention that here.

File: README.md [hunk 7]
```diff
@@ -2,51 +2,126 @@
...
+  TODO: what's a test configuration file?
+
+* `lua_libs` — paths to Lua files that should be copied to the directory,
+  where a server is started. For example:
+
+  ```ini
+  lua_libs = lua/require_mod.lua lua/serializer_test.lua lua/process_timeout.lua
+  ```
+
+### Skipping tests
```
Comment (Totktonada) [inline README.md:54]: I suggest to avoid using two terms for the same entity. It is confusing per se, but also it clashes with skipping of tests using `*.skipcond` files. Disabled and skipped tests are reported differently by test-run. (We can argue, whether it is good or bad, but it is how test-run present information of the screen. So readme should reflect it to don't introduce a confusion.)

File: README.md [hunk 8]
```diff
@@ -2,51 +2,126 @@
...
+
+A number of fields are used to disable (skip) certain tests:
+
+* `disabled` — list of tests that should be skipped
+* `release_disabled` — list of tests that should be skipped when Tarantool is
```
Comment (Totktonada) [inline README.md:59]: Worth to add: usually tests, which use error injections (which are available only in debug tarantool builds).

File: README.md [hunk 9]
```diff
@@ -2,51 +2,126 @@
...
+    built with `Release`
```
Comment (Totktonada) [inline README.md:60]: Strictly speaking, tarantool is built either with `Debug` or with `RelWithDebInfo` build type.

File: README.md [hunk 10]
```diff
@@ -2,51 +2,126 @@
...
+* `valgrind_disabled` — list of tests that should be skipped when Valgrind is enabled
+* `long_run` — mark tests as long, such tests will run only with `--long` option.
```
Comment (Totktonada) [inline README.md:62]: It would be good to give a suggestions, what is considered 'short' and what 'long'. A long time ago the following informal rule was spoken: a test should usually pass within 1 second. Now it is far away of truth. Maybe not a question for this PR.

File: README.md [hunk 11]
```diff
@@ -2,51 +2,126 @@
...
+
+  ```ini
+  long_run = t1.test.lua t2.test.lua
+  ```
+
+### Other parameters
+
+
+* `show_reproduce_content` (optional, `True` by default) — TODO: what's this?
```
Comment (Totktonada) [inline README.md:71]: https://github.com/tarantool/test-run/issues/122

File: README.md [hunk 12]
```diff
@@ -2,51 +2,126 @@
...
+* `is_parallel` (optional, `False` by default) — whether the tests in the suite can run in parallel
+
+* `use_unix_sockets` (optional, `False` by default) — use hard-coded UNIX sockets
```
Comment (Totktonada) [inline README.md:74]: For console, AFAIR.

'Hard-coded' is a bit too hard words. It depends of server's script name.

File: README.md [hunk 13]
```diff
@@ -2,51 +2,126 @@
...
+and an optional skip condition file `*.skipcond`.
+
+In the beginning of each test run, the `.skipcond` file is executed as a Python script.
+In the local Python environment there's a `self` object, which is an instance of the [`Test` class](./lib/test.py).
+To skip the test under some condition, set `self.skip = 1`.
+
+Next, the test in `.test(.lua|.py)?` is executed. 
+If this is a first test run and there's no `.result` file yet, it is generated from the output.
+If there is a `.result` file, output is written to a `.reject` file,
+which is then compared to `.result`.
```
Comment (Totktonada) [inline README.md:110]: 1. `.test.sql` as well.
2. And `_test.lua` in luatest test suites.
3. The autowritting `.result` file was removed together with introducing the `--update-result` option.

File: README.md [hunk 14]
```diff
@@ -2,51 +2,126 @@
...
+
+If there's a difference between `.reject` and `.result`, the following happens:
+
+1. The last 15 lines of diff are printed to output.
+2. The `.reject` file is saved in the `<vardir>/rejects/<suite>`
+   subfolder given in options or set locally as `var/rejects/<suite>` by default.
+
+If there is no difference between `.reject` and `.result`, the `.reject` file is deleted.
```
Comment (Totktonada) [inline README.md:118]: Don't know, whether it is useful, to be honest.

File: README.md (hunk 13 above)
Comment (Totktonada) [reply README.md:110]: 4. There is no a word about TAP13 compatible output (it does not need a result file).

File: README.md (hunk 11 above)
Comment (NickVolynkin) [reply README.md:71]: thanks, I've added this info to the document

File: README.md [hunk 15]
```diff
@@ -2,51 +2,185 @@
...
+}
+```
+
+## Test composition
+
+Each test consists of the following files:
+
+* Test file: `<name>.test.lua`, `<name>_test.lua`, `<name>.test.py`, or `<name>.test.sql`.
+* Reference file: `<name>.result`.
+* Skip condition file: `<name>.skipcond`.
```
Comment (Totktonada) [inline README.md:150]: The altter two bullets are optional, but the text looks like they're mandatory.

File: README.md [hunk 16]
```diff
@@ -62,15 +196,15 @@ configuration):
 }
 ```
 
-In test case we can get configuration from inspector:
+In the test case we can get configuration from the inspector:
 
 ```lua
 engine = test_run:get_cfg('engine')
 -- first run engine is 'memtx'
 -- second run engine is 'sophia'
```
Comment (Totktonada) [inline README.md:204]: Let's replace sophia with vinyl.

File: README.md [hunk 17]
```diff
@@ -62,15 +196,15 @@ configuration):
...
 ```
 
-"engine" value has a special meaning for *.test.sql files: if it is "memtx" or
+"engine" value has a special meaning for `*.test.sql` files: if it is "memtx" or
 "vinyl", then the corresponding default engine will be set before executing
 commands from a test file. An engine is set with the following commands:
```
Comment (Totktonada) [inline README.md:209]: Regarding commands below: maybe it is too deep details for a user guide.

File: README.md [hunk 18]
```diff
@@ -79,11 +213,13 @@ UPDATE "_session_settings" SET "value" = 'memtx|vinyl' WHERE "name" = 'sql_defau
...
+
+## Writing various types of tests
 
-#### Python
+### Python tests
 
-Files: `<name>.test.py`, `<name>.result` and `<name>.skipcond`(optionaly).
+Files: `<name>.test.py`, `<name>.result` and `<name>.skipcond` (optionally).
 
 Environment:
```
[the remark this answers is not in the document]
Comment (Totktonada) [reply README.md:224]: While `admin` below still actual.

[the remark this answers is not in the document]
Comment (Totktonada) [reply README.md:224]: I also suggest to re-verify that the example below still work.

File: README.md [hunk 19]
```diff
@@ -166,14 +302,14 @@ box.info.lsn
 ...
 ```
 
-#### Lua
+### Lua
 
 Files: `<name>.test.lua`, `<name>.result` and `<name>.skipcond`(optionaly).
 Tests interact only with `AdminConnection`. Supports some preprocessor functions (eg `delimiter`)
```
Comment (Totktonada) [inline README.md:308]: > Tests interact only with `AdminConnection`.

Don't know what it means, but it looks irrelevant now.

File: README.md [hunk 20]
```diff
@@ -166,14 +302,14 @@ box.info.lsn
 ...
 ```
 
-#### Lua
+### Lua
```
Comment (Totktonada) [inline README.md:305]: `core = tarantool` and `core = app` tests are really different. It worth to split its description.

File: README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (NickVolynkin) [reply README.md:224]: I don't want to touch this part of readme in this PR. Will review it later.

File: README.md [hunk 21]
```diff
@@ -2,51 +2,185 @@
...
+Whenever a test fails, the `.reject` file is saved in the `<vardir>/rejects/<suite>`
+subdirectory given in options or set locally as `var/rejects/<suite>` by default.
+If the test is considered successful, the `.reject` file is deleted.
+
+## Test configuration
+
+Test configuration file contains configuration for multiple runs.
+For each test section, system runs a separate test and compares the result to the common `.result` file.
+For  example, we need to run a test with different DB engines (`"*"` means the default
 configuration):
```
[the remark this answers is not in the document]
Comment (NickVolynkin) [reply README.md:184]: In this example

```json
{
    "first.test.lua": {
      "memtx": {"engine": "memtx"}
    },
    "second.test.lua": {},
    "*": {
        "memtx": {"engine": "memtx"},
        "vinyl": {"engine": "vinyl"}
    }
}
```

what's the difference between first and second tests?

**Comment by Totktonada** [comment]
My comments are mostly not answered. It is in the responsibility of a developer to update all discussions before asking for a next review. Otherwise a reviewer will be obligated to re-check every point, which (s)he highlighted. It is time consuming and not profitable.

I see a temporary commit in this pull request. I expect ready-to-go (in your opinion) pull request, when a review is acquired. Please, elaborate.

After this I'll look at your answers and, if nothing will confuse me, can skip re-check of the changes. You'll wait less time for review so.

File: README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply README.md]: Both are run once, but the first on will receive `'memtx'` from a `test_run:get_cfg('engine')` call. A test can be writtten in any way, the difference is mostly how the configuration information is presented for a reader.

File: README.md (hunk 1 above)
Comment (NickVolynkin) [reply README.md:48]: removed this part

File: README.md (hunk 11 above)
Comment (NickVolynkin) [reply README.md]: I've found a single use case: `./sql-tap/suite.ini:31:show_reproduce_content = False`. So, for all other tests this value is `True` (by default), which is useless — is that really so?

File: README.md (hunk 13 above)
Comment (NickVolynkin) [reply README.md]: I've removed enumeration of test file extensions from this line, it resolves (1) and (2).  Also, I won't mention writing `.result` files at all. Moved it to a separate issue: https://github.com/tarantool/test-run/issues/327

Added a few words about TAP13

File: README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply README.md:242]: How I'll differentiate parts, where problems are left deliberately and where it is unintentional? No, I think your point does not work here: your changes are not so dense.

**Comment by Totktonada** [comment]
> My comments are mostly not answered.

Still. Please, find another reviewer.

**Comment by Totktonada** [comment]
Seems like nobody is eager to work on it.
